### ISSUE TITLE
next method of FederatedDataLoader takes forever on remote workers

### ISSUE THREAD

**Issue by danielhomola** [OpenMined/PySyft#2522]
I took the MNIST (the basic, non-parallel version) tutorial from the advanced tutorials and replaced the local WebsocketClientWorker's with 3 identical WebsocketClientWorker on an EC2 instance with only CPU (no GPU acceleration). 

Connecting, asking the workers how are they (various props, methods) works like a charm and even the full tutorial can be completed (i.e. distributing the MNIST dataset to them and training on AWS, then predicting the test set locally). However, it took 4 hours. 

I was curious, so played around with parallelism, batch size, etc, monitored everything in htop, but then finally I had to get to log every line and IPython embed here and there to work out, that the bottleneck is [here](https://github.com/OpenMined/PySyft/blob/54a141694d0e63f1b812aeda91a10efbf4f83149/examples/tutorials/advanced/websockets-example-MNIST/run_websocket_client.py#L107), i.e. when calling the __next__ of the FederatedDataLoader iterator. 

It takes (regardless of the batch size) ages, as it only uses 2% of the 8 cores on the EC2 machine. Once the training function gets to actual training (i..e PyTorch stuff) it's great and spins up all cores, maxes them out, etc.. 

Any idea, why this could be? Usually batch prep is done on CPU (at least most commonly in TF - I'm new to Torch). So even if I put a GPU to this, it should stay the same.. 

Thanks very much for any hint,
d

**Comment by jvmancuso** [comment]
If I had to guess, I would say this is what the original DataLoader's `num_workers` solves.  It makes sense to prepare data in parallel, so that you always have a data point ready, but based on your feedback this is not the case for FederatedDataLoader just yet.  I suspect that the communication required to prepare a single batch of data is pretty costly as well, so a Plan can be used to improve that.  We would love to see a deep dive into the codebase around this issue, I'm sure there are some quick wins to be had!
